### ISSUE TITLE
An instanceof check in a switch(true) case should refine the variable type

### ISSUE THREAD

**Issue by stof** [scrutinizer-ci/php-analyzer#526]
In several libraries I know of (Behat and Doctrine for instance), ``switch (true)`` is used instead of a bunch of ``elseif`` statement. However, PhpAnalyzer does not support refining the type of variables based on conditions like that, thus reporting false positives:

```php
switch (true) {
    case ($context instanceof ScenarioContext):
        // $context should be considered as being a ScenarioContext here
        break;
}

```

see https://scrutinizer-ci.com/g/Behat/Behat/inspections/8322aa43-d58e-4659-8ba4-25180f60d61a/issues/files/src/Behat/Behat/EventDispatcher/Event/GherkinEventFactory.php?status=new&orderField=path&order=asc&issueId=6ea06f55-7402-492e-b607-0feb06104b83

**Comment by schmittjoh** [comment]
There was a problem with short-circuiting expressions like ``&&``, and ``||``; this is fixed now.
